### ISSUE TITLE
error in rsync protocol data stream (code 12) at io.c(788)

### ISSUE THREAD

**Issue by StevenP94** [backuppc/backuppc#116]
Hello, I just updated from v3 and can't backup anything due to this error.
I deleted the old v3 topdir and installed a new one so this should be a pure v4 installation (I've update config.pl too).
This is the xfer log (loglevel 7):
```
XferLOG file /var/lib/backuppc/pc/stefano/XferLOG.0.z created 2017-05-16 07:28:37 
Backup prep: type = full, case = 1, inPlace = 1, doDuplicate = 0, newBkupNum = 0, newBkupIdx = 0, lastBkupNum = , lastBkupIdx =  (FillCycle = 0, noFillCnt = )
Running: /usr/local/bin/rsync_bpc --bpc-top-dir /var/lib/backuppc --bpc-host-name stefano --bpc-share-name docs --bpc-bkup-num 0 --bpc-bkup-comp 3 --bpc-bkup-prevnum -1 --bpc-bkup-prevcomp -1 --bpc-bkup-inode0 1 --bpc-attrib-new --bpc-log-level 7 --super --recursive --protect-args --numeric-ids --perms --owner --group -D --times --links --hard-links --delete --delete-excluded --one-file-system --partial --log-format=log:\ %o\ %i\ %B\ %8U,%8G\ %9l\ %f%L --stats --checksum --iconv=utf8,cp1252 --timeout=72000 --password-file=/var/lib/backuppc/pc/stefano/.rsyncdpw29018 backuppc@192.168.0.1::docs /
full backup started for directory docs
Xfer PIDs are now 29033
This is the rsync child about to exec /usr/local/bin/rsync_bpc
bpc_path_create(/var/lib/backuppc/pc/stefano/0)
bpc_attrib_backwardCompat: WriteOldStyleAttribFile = 0, KeepOldAttribFiles = 0
rsync_bpc: read error: Connection reset by peer (104)
bpc_sysCall_cleanup: doneInit = 1
RefCnt Deltas for new backup
Uncompressed HT:
Compressed HT:
Done: 0 errors, 0 filesExist, 0 sizeExist, 0 sizeExistComp, 0 filesTotal, 0 sizeTotal, 0 filesNew, 0 sizeNew, 0 sizeNewComp, 1 inode
Parsing done: nFilesTotal = 0
rsync error: error in rsync protocol data stream (code 12) at io.c(788) [Receiver=3.0.9.6]
rsync_bpc exited with fatal status 12 (3072) (rsync error: error in rsync protocol data stream (code 12) at io.c(788) [Receiver=3.0.9.6])
Xfer PIDs are now 
Got fatal error during xfer (No files dumped for share docs)
Backup aborted (No files dumped for share docs)
BackupFailCleanup: nFilesTotal = 0, type = full, BackupCase = 1, inPlace = 1, lastBkupNum = 
Removing empty backup #0
Xfer PIDs are now 29038
BackupPC_backupDelete: removing #0
BackupPC_backupDelete: No prior backup for merge
Xfer PIDs are now 29039,29038
BackupPC_refCountUpdate: computing totals for host stefano
bpc_attrib_backwardCompat: WriteOldStyleAttribFile = 0, KeepOldAttribFiles = 0
BackupPC_refCountUpdate: host stefano got 0 errors (took 0 secs)
BackupPC_refCountUpdate total errors: 0
Xfer PIDs are now 29038
BackupPC_backupDelete: got 0 errors
Xfer PIDs are now 
Finished BackupPC_backupDelete, status = 0 (running time: 0 sec)
Xfer PIDs are now 
Running BackupPC_refCountUpdate -h stefano -f on stefano
Xfer PIDs are now 29040
BackupPC_refCountUpdate: computing totals for host stefano
bpc_attrib_backwardCompat: WriteOldStyleAttribFile = 0, KeepOldAttribFiles = 0
BackupPC_refCountUpdate: host stefano got 0 errors (took 0 secs)
BackupPC_refCountUpdate total errors: 0
Xfer PIDs are now 
Finished BackupPC_refCountUpdate (running time: 0 sec)
Xfer PIDs are now 
```
Any hint ? I have the same on all 3 clients (all running Windows 10 with rsyncd)
Thanks in advance

**Comment by craigbarratt** [comment]
First, issues on github are more meant for potential bugs rather than support questions.  You should use the user email list instead.

Anyhow, for some reason the Win10 client shuts down the socket connection before it gets very far.  I'd recommend you add a "-vvv" option to $Conf{RsyncArgs} and try again.

Is there any chance that your backup host IP address has changed, and your clients are configured to only accept connections from a different (original) address?

**Comment by StevenP94** [comment]
Thanks for your reply, and excuse me if I submitted a support request here, btw all IPs (clients and server) are fixed.

This is the result after adding -vvv:

XferLOG file /var/lib/backuppc/pc/stefano/XferLOG.0.z created 2017-05-16 08:36:25 
Backup prep: type = full, case = 1, inPlace = 1, doDuplicate = 0, newBkupNum = 0, newBkupIdx = 0, lastBkupNum = , lastBkupIdx =  (FillCycle = 0, noFillCnt = )
Running: /usr/local/bin/rsync_bpc --bpc-top-dir /var/lib/backuppc --bpc-host-name stefano --bpc-share-name docs --bpc-bkup-num 0 --bpc-bkup-comp 3 --bpc-bkup-prevnum -1 --bpc-bkup-prevcomp -1 --bpc-bkup-inode0 1 --bpc-attrib-new --bpc-log-level 7 --super --recursive --protect-args --numeric-ids --perms --owner --group -D --times --links --hard-links --delete --delete-excluded --one-file-system --partial --log-format=log:\ %o\ %i\ %B\ %8U,%8G\ %9l\ %f%L --stats -vvv --checksum --iconv=utf8,cp1252 --timeout=72000 --password-file=/var/lib/backuppc/pc/stefano/.rsyncdpw29772 backuppc@192.168.0.1::docs /
full backup started for directory docs
Xfer PIDs are now 29787
This is the rsync child about to exec /usr/local/bin/rsync_bpc
bpc_path_create(/var/lib/backuppc/pc/stefano/0)
bpc_attrib_backwardCompat: WriteOldStyleAttribFile = 0, KeepOldAttribFiles = 0
opening tcp connection to 192.168.0.1 port 873
Connected to 192.168.0.1 (192.168.0.1)
client charset: utf8
sending daemon args: --server --sender -svvvlHogDtprcxe.iLsf --iconv=utf8 
rsync_bpc: read error: Connection reset by peer (104)
bpc_sysCall_cleanup: doneInit = 1
RefCnt Deltas for new backup
Uncompressed HT:
Compressed HT:
Done: 0 errors, 0 filesExist, 0 sizeExist, 0 sizeExistComp, 0 filesTotal, 0 sizeTotal, 0 filesNew, 0 sizeNew, 0 sizeNewComp, 1 inode
Parsing done: nFilesTotal = 0
rsync error: error in rsync protocol data stream (code 12) at io.c(788) [Receiver=3.0.9.6]
[Receiver] _exit_cleanup(code=12, file=io.c, line=788): about to call exit(12)
rsync_bpc exited with fatal status 12 (3072) (rsync error: error in rsync protocol data stream (code 12) at io.c(788) [Receiver=3.0.9.6])

please can you give me a link for the support email list ?
thanks again

**Comment by craigbarratt** [comment]
What happens if you do:

telnet 192.168.0.1 873

The user list is backuppc-users@lists.sourceforge.net.

Also, could you try removing the --one-file-system option?

Craig

**Comment by drenton1** [comment]
What happens if you remove cp1252 from ClientCharset?

Dale

**Comment by StevenP94** [comment]
@craigbarratt:

telnet 192.168.0.1 873:
Trying 192.168.0.1...
Connected to 192.168.0.1.
Escape character is '^]'.
@RSYNCD: 30.0

removing --one-file-system gives the same result:
XferLOG file /var/lib/backuppc/pc/stefano/XferLOG.0.z created 2017-05-16 08:36:25 
Backup prep: type = full, case = 1, inPlace = 1, doDuplicate = 0, newBkupNum = 0, newBkupIdx = 0, lastBkupNum = , lastBkupIdx =  (FillCycle = 0, noFillCnt = )
Running: /usr/local/bin/rsync_bpc --bpc-top-dir /var/lib/backuppc --bpc-host-name stefano --bpc-share-name docs --bpc-bkup-num 0 --bpc-bkup-comp 3 --bpc-bkup-prevnum -1 --bpc-bkup-prevcomp -1 --bpc-bkup-inode0 1 --bpc-attrib-new --bpc-log-level 7 --super --recursive --protect-args --numeric-ids --perms --owner --group -D --times --links --hard-links --delete --delete-excluded --one-file-system --partial --log-format=log:\ %o\ %i\ %B\ %8U,%8G\ %9l\ %f%L --stats -vvv --checksum --iconv=utf8,cp1252 --timeout=72000 --password-file=/var/lib/backuppc/pc/stefano/.rsyncdpw29772 backuppc@192.168.0.1::docs /
full backup started for directory docs
Xfer PIDs are now 29787
This is the rsync child about to exec /usr/local/bin/rsync_bpc
bpc_path_create(/var/lib/backuppc/pc/stefano/0)
bpc_attrib_backwardCompat: WriteOldStyleAttribFile = 0, KeepOldAttribFiles = 0
opening tcp connection to 192.168.0.1 port 873
Connected to 192.168.0.1 (192.168.0.1)
client charset: utf8
sending daemon args: --server --sender -svvvlHogDtprcxe.iLsf --iconv=utf8 
rsync_bpc: read error: Connection reset by peer (104)
bpc_sysCall_cleanup: doneInit = 1
RefCnt Deltas for new backup
Uncompressed HT:
Compressed HT:
Done: 0 errors, 0 filesExist, 0 sizeExist, 0 sizeExistComp, 0 filesTotal, 0 sizeTotal, 0 filesNew, 0 sizeNew, 0 sizeNewComp, 1 inode
Parsing done: nFilesTotal = 0
rsync error: error in rsync protocol data stream (code 12) at io.c(788) [Receiver=3.0.9.6]
[Receiver] _exit_cleanup(code=12, file=io.c, line=788): about to call exit(12)
rsync_bpc exited with fatal status 12 (3072) (rsync error: error in rsync protocol data stream (code 12) at io.c(788) [Receiver=3.0.9.6])

**Comment by StevenP94** [comment]
@drenton1 removing cp1252 from ClientCharset has no effect, using utf8 has no effect too, the log is the same and reports the same problem.

**Comment by StevenP94** [comment]
Update: using DeltaCopy as client on Windows 10 instead of rsyncd version 3.0.9, seems to work but need to remove the -iconv option that is unsupported.
Any idea on how to do that ?

**Comment by craigbarratt** [comment]
If you set $Conf{ClientCharset} to "" (empty string), then the iconv option
to rsync will not be used.

Craig

**Comment by StevenP94** [comment]
@craigbarratt  Solved!
I installed cygwin-rsyncd-3.1.2.0 and finally the error was logged on client:
rsync: The server is configured to refuse --iconv
digging on internet I solved adding this to client configuration rsyncd.conf:
charset = utf-8
I restarted the client daemon and then backuppc start to works and I completed my backup.
This problem was really hard to solve because the client 3.0.9 has absolutely no info on log, just the connect received, and there was no info or error message to help to understand what happens.
I think that all windows clients using cygwin-rsyncd-3.0.9.0 can have this problem, maybe the README should be updated and this info should be included.
Btw thanks for your support and excuse me for my english
